src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs

```diff
@@ -233,16 +236,35 @@ private BoundForEachStatement BindForEachPartsWorker(DiagnosticBag diagnostics,
                         iterationVariableType = inferredType ?? CreateErrorType("var");
 
                         var variables = node.Variable;
-                        var valuePlaceholder = new BoundDeconstructValuePlaceholder(_syntax.Expression, iterationVariableType);
-                        BoundDeconstructionAssignmentOperator deconstruction = BindDeconstruction(
-                                                                                variables,
-                                                                                variables,
-                                                                                right: null,
-                                                                                diagnostics: diagnostics,
-                                                                                isDeclaration: true,
-                                                                                rightPlaceholder: valuePlaceholder);
-
-                        deconstructStep = new BoundForEachDeconstructStep(variables, deconstruction, valuePlaceholder);
+                        if (variables.Kind() == SyntaxKind.TupleExpression || variables.Kind() == SyntaxKind.DeclarationExpression)
+                        {
+                            var valuePlaceholder = new BoundDeconstructValuePlaceholder(_syntax.Expression, iterationVariableType);
+                            CSharpSyntaxNode declaration = null;
+                            CSharpSyntaxNode expression = null;
+                            BoundDeconstructionAssignmentOperator deconstruction = BindDeconstruction(
+                                                                                    variables,
+                                                                                    variables,
+                                                                                    right: null,
+                                                                                    diagnostics: diagnostics,
+                                                                                    rightPlaceholder: valuePlaceholder,
+                                                                                    declaration: ref declaration,
+                                                                                    expression: ref expression);
+
+                            if (expression != null)
+                            {
+                                // error: must declare foreach loop iteration variables.
+                                Error(diagnostics, ErrorCode.ERR_MustDeclareForeachIteration, variables);
+                                hasErrors = true;
+                            }
+
+                            deconstructStep = new BoundForEachDeconstructStep(variables, deconstruction, valuePlaceholder);
+                        }
+                        else if (!node.HasErrors)
```

Comment:
`CodeGenDeconstructTests.ShortDiscardDisallowedInForeach`  is the only way this can occur (i.e. a simple name `_`).

---
In reply to: [91418538](https://github.com/dotnet/roslyn/pull/15548#discussion_r91418538) [](ancestors = 91418538,90178336,90174311)